server/src/main/java/org/candlepin/model/dto/ProductData.java

```diff
@@ -1195,9 +1206,19 @@ public ProductData populate(Product source) {
         this.setBranding(source.getBranding());
 
         if (source.getProvidedProducts() != null) {
-            this.setProvidedProducts(source.getProvidedProducts().stream()
-                .map(prod -> new ProductData(prod))
-                .collect(Collectors.toSet()));
+            if (this.providedProducts == null) {
+                this.providedProducts = new HashSet<>();
+            }
+            else {
+                this.providedProducts.clear();
+            }
+
+            for (Product pData : source.getProvidedProducts()) {
+                this.providedProducts.add(new ProductData(pData));
+            }
+        }
+        else {
+            this.setProvidedProducts(null);
```

Comment:
This is done to avoid creating a intermediate collection while collecting all providedProducts within a Java stream as suggested by @Ceiu  
The method `this.setProvidedProducts` doesn't allow deep copy which is needed here in populate method.